lib/mempool/mempool.js

```diff
@@ -1851,6 +1885,75 @@ class Mempool extends EventEmitter {
     }
   }
 
+  /**
+   * Recursively disconnect spenders of a transaction.
+   * @private
+   * @param {TX} parent
+   */
+
+  disconnectSpenders(parent) {
+    const parentHash = parent.hash();
+    const children = new BufferSet();
+    const names = new BufferSet();
+    rules.addNames(parent, names);
+
+    for (let i = 0; i < parent.outputs.length; i++) {
+      const spender = this.getSpent(parentHash, i); // MempoolEntry
+
+      // No child spending this output
+      if (!spender)
+        continue;
+
+      const {tx} = spender;
+
+      // Child is not linked by name.
+      // Covenant rules don't prevent it from
+      // staying in the mempool.
+      if (!rules.hasNames(tx, names))
+        continue;
+
+      // This child has to be removed from the mempool
+      // but it will be valid again once its parent confirms.
+      this.disconnectSpenders(tx);
+      children.add(tx.hash());
+      this.children.set(tx.hash(), tx);
+      this.removeEntry(spender);
+    }
+
+    if (children.size)
+      this.disconnected.set(parentHash, children);
+  }
+
+  /**
+   * Recursively reconnect spenders of a transaction.
+   * @private
+   * @param {TX} parent
+   */
+
+  async reconnectSpenders(parent) {
+    const hashes = this.disconnected.get(parent.hash());
+
+    if (!hashes)
+      return;
+
+    for (const hash of hashes) {
+      const tx = this.children.get(hash);
+      assert(tx);
```

Comment:
Note: this assert will no longer hold true if we track double spends for children.